lib/topology/pytest/plugin.py

```diff
@@ -352,9 +421,221 @@ def pytest_runtest_setup(item):
+        # this is a new module and we need to calculate the topology_hash
+        module = item.module
+        module_name = item.module.__name__
+        topology_hash = next(
+            (
+                key for key, value in unique_topologies.items()
+                if module_name in value['modules']
+            ),
+            None
+        )
+
+        if topology_hash is None:
+
+            # This is a new topology, so calculate a new topology_hash.
+            # To do that, we need to obtain the topology as a dict and merge
+            # it with the injected attributes if any. Once merged, we can
+            # calculate the topology_hash using DeepHash.
+            topology, injected_attr = get_module_topology(plugin, module)
+
+            if injected_attr:
+
+                inject_nodes = injected_attr.get('nodes', {})
+                inject_ports = injected_attr.get('ports', {})
+                inject_links = injected_attr.get('links', {})
+
+                # Merge the topology nodes with the inject nodes
+                for topo_node in topology.get('nodes', []):
+                    for node_id in topo_node.get('nodes', []):
+                        if node_id in inject_nodes:
+                            topo_node['attributes'].update(
+                                inject_nodes[node_id]
+                            )
+
+                # Merge the topology ports with the inject ports
+                for topo_port in topology.get('ports', []):
+                    for node_id, port in topo_port.get('ports', []):
+                        if (node_id, port) in inject_nodes:
+                            topo_port['attributes'].update(
+                                inject_ports[(node_id, port)]
+                            )
+
+                # Merge the topology links with the inject links
+                for topo_link in topology.get('links', []):
+                    if topo_link['endpoints'] in inject_links:
+                        topo_link['attributes'].update(
+                            inject_links[topo_link['endpoints']]
+                        )
+
+            # Find the topology_hash node or create a new one and add the
+            # module to the the node
+            # From deepdiff docs:
+            #   At first it might seem weird why DeepHash(obj)[obj] but
+            #   remember that DeepHash(obj) is a dictionary of hashes of all
+            #   other objects that obj contains too.
+            # Reference:
+            #   https://deepdiff.readthedocs.io/en/latest/deephash.html
+            topology_hash = DeepHash(topology)[topology]
+            module.__TOPOLOGY_HASH__ = topology_hash
+            node = unique_topologies.setdefault(topology_hash, {
+                'items': [],
+                'modules': {},
+                'topology': topology,
+                'attributes': ''
+            })
+            node['modules'][module_name] = []
```

Comment:
I think "node" is not a good name here, maybe "module_topology" or "unique_topology"?